### PR TITLE
Add nix flake

### PR DESCRIPTION
**Pull request by kanielrkirby**
- Updated `README.md` with Nix installation options.
- Added a `flake.nix` and `flake.lock` for flake-compatible setups.
- Added a `default.nix` using `flake-compat` for compatibility to non-flake setups.

Addresses #35.

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    flake-parts.url = "github:hercules-ci/flake-parts";
+    naersk.url = "github:nix-community/naersk";
+    flake-compat.url = "https://flakehub.com/f/edolstra/flake-compat/1.tar.gz";
+  };
+  outputs = { flake-parts, naersk, nixpkgs, ... }@inputs:
+    flake-parts.lib.mkFlake { inherit inputs; } {
+      systems = [
+        "x86_64-linux"
+        "x86_64-darwin"
+        "aarch64-darwin"
```
Comment (koalazub) [inline flake.nix:13]: We should probably add "aarch64-linux" as well

**Review by koalazub — APPROVED** [review]

**Review by MilesCranmer — COMMENTED** [review]
Is there a way you can add a test for this in the GitHub action?

File: README.md [hunk 2]
```diff
@@ -48,6 +48,56 @@ made available on the GitHub releases page: https://github.com/MilesCranmer/rip2
 
 To install, simply open the archive and move the binary somewhere you can run it.
 
+### Nix
+
+This repository is flake-compatible, and backwards-compatible with non-flake systems using `flake-compat`. It uses `naersk` to build the Rust package from source.
+
+#### Try Out Before Installing (With Flakes)
```
Comment (MilesCranmer) [inline README.md:55]: Could we wrap the detailed instructions with a toggle?

```
<details>
<summary>
…
</summary>

(Full installation instructions)

</details>
```

Then it won’t take up as much room in the readme (see cargo instructions for comparison)

Comment (MilesCranmer) [reply README.md:55]: See https://gist.github.com/Ashton-W/864d09d71f7cb8e5b47ebaac02b98455

File: flake.nix (hunk 1 above)
Comment (kanielrkirby) [reply flake.nix:14]: Added that in the recent commit

File: README.md (hunk 2 above)
Comment (kanielrkirby) [reply README.md:56]: Added that in the recent commit

**Comment by kanielrkirby** [comment]
Added a test in the GH workflow, let me know if it needs tweaking (as I'm very very not good at workflows lol)

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -56,6 +56,15 @@ jobs:
         run: cargo doc --all --no-deps
       - name: Test with code coverage
         run: cargo tarpaulin --release --engine llvm --follow-exec --post-test-delay 10 --coveralls ${{ secrets.COVERALLS_REPO_TOKEN }}
+<<<<<<< Updated upstream
+=======
```
Comment (koalazub) [inline .github/workflows/ci.yml:60]: think these were supposed to be deleted?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -56,6 +56,15 @@ jobs:
...
+      - name: Check Nix flake inputs
+        uses: DeterminateSystems/flake-checker-action@v4
+      - name: Install Nix
+        uses: DeterminateSystems/nix-installer-action@v3
+      - name: Build default package
+        run: nix build
+>>>>>>> Stashed changes
```
Comment (koalazub) [inline .github/workflows/ci.yml:67]: and here?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -56,6 +56,15 @@ jobs:
...
       - name: Test with code coverage
         run: cargo tarpaulin --release --engine llvm --follow-exec --post-test-delay 10 --coveralls ${{ secrets.COVERALLS_REPO_TOKEN }}
+<<<<<<< Updated upstream
+=======
+      - name: Check Nix flake inputs
+        uses: DeterminateSystems/flake-checker-action@v4
+      - name: Install Nix
+        uses: DeterminateSystems/nix-installer-action@v3
+      - name: Build default package
+        run: nix build
```
Comment (koalazub) [inline .github/workflows/ci.yml:66]:
```suggestion
        run: nix build --dry-run
```

Only if you're testing and not needing anything else

**Review by koalazub — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 5 above)
Comment (kanielrkirby) [reply .github/workflows/ci.yml:66]: Yup, my bad, always got the Git issues lol

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -56,6 +56,12 @@ jobs:
         run: cargo doc --all --no-deps
       - name: Test with code coverage
         run: cargo tarpaulin --release --engine llvm --follow-exec --post-test-delay 10 --coveralls ${{ secrets.COVERALLS_REPO_TOKEN }}
+      - name: Check Nix flake inputs
```
Comment (MilesCranmer) [inline .github/workflows/ci.yml:59]: Can this be added in a separate, parallel, job, within ci.yml? Also need to have a test for whether the nix-built `rip` works

File: README.md [hunk 7]
```diff
@@ -48,6 +48,59 @@ made available on the GitHub releases page: https://github.com/MilesCranmer/rip2
 
 To install, simply open the archive and move the binary somewhere you can run it.
 
+<details>
+<summary>Nix</summary>
```
Comment (MilesCranmer) [inline README.md:52]:
```suggestion
<summary>Additional nix options</summary>
```

And then add a normal header/description before the `<details>` block, similar to what is done for Cargo and the binaries. Like

```
### Nix flake

(Short description, maybe show the shell command example for quickly running it)

(then, the further details wrapped in a toggle)
```

File: shell.nix [hunk 8]
```diff
@@ -0,0 +1,10 @@
+(import
```
Comment (MilesCranmer) [inline shell.nix:1]: Not sure if it’s possible but is there any way that `shell.nix` and `default.nix` can instead just be included within `flake.nix`? Or do they really need to be three separate files?

**Review by MilesCranmer — CHANGES_REQUESTED** [review]

File: shell.nix (hunk 8 above)
Comment (kanielrkirby) [reply shell.nix:1]: Unfortunately, the community is pretty divided on flakes, and there's not much integration between the people using them and the people not using them. The flake is the modern way to do things, and makes things significantly easier, while the `default` is the old way of doing things, using the flake as a source of truth.

However, I *can* move everything (or at the very least `default`, haven't tested past that before) to a `nix` directory, losing some ergonomics, as well as put the logic for the shell in the `default.nix`. Not sure why I had made the shell instead of doing that, I think I was confused because I'm always confused while using Nix haha.

TL;DR, I'll move the shell into the `default.nix` at least, do you want me to try moving everything in a `nix` directory? I'm not sure the `flake.nix` *can*, but I can try, I know the `default.nix` can.

Comment (MilesCranmer) [reply shell.nix:1]: Oh I see, no worries then! Let's just keep them as .nix files in the root directory.

Comment (kanielrkirby) [reply shell.nix:1]: Update: `flake.nix` can, but it makes using it as a Nix flake a bit more convoluted / less expected (e.g., instead of `nix run "github:MilesCranmer/rip2"` it'll be `nix run "github:MilesCranmer/rip2?dir=nix"`). It's a usability concern at most though, as far as I can tell. Up to y'all, what do you think?

Comment (MilesCranmer) [reply shell.nix:1]: I think I would just go with what most people do. I'm assuming people will just see there is a .nix file here and then run the command they always run on the github URL.

**Comment by kanielrkirby** [comment]
@MilesCranmer Should all be good, let me know if there's any other fixes! Learned a lot about how Nix files are integrated in real world scenarios here, now I can add them to my own! :p

**Comment by kanielrkirby** [comment]
Also am sorry for that horribly named commit message, that's my bad. I could have sworn I split them up haha.

**Comment by MilesCranmer** [comment]
Seems like it's not being installed to that dir?

**Comment by MilesCranmer** [comment]
Also I just realised something else. 

> * Added a `default.nix` using `flake-compat` for compatibility to non-flake setups.

It sounds like `default.nix` is a different build file entirely than the flake one? If so we should probably have another test for that (or just delete it)

**Comment by kanielrkirby** [comment]
> Also I just realised something else.
> 
> > * Added a `default.nix` using `flake-compat` for compatibility to non-flake setups.
> 
> It sounds like `default.nix` is a different build file entirely than the flake one? If so we should probably have another test for that (or just delete it)

Happy to add a test for it if you still think it's necessary, but two things first:

1. `default.nix` pulls it's source of truth completely from the flake through a compatibility layer (`flake-compat`), so if it works, there's almost no world where `flake.nix` doesn't work, since it's just iterating through every node in the `flake.lock` / `flake.nix`, and manually handling it.
2. If the build wasn't going to work for whatever reason for the `flake.nix`, the `flake-checker` would fail.

Definitely can still add a test for it separately if you think it's worth it, just wanted to check first.

**Comment by kanielrkirby** [comment]
> Seems like it's not being installed to that dir?

Also interesting, I'll have to look into that, that's where it should be building to.

**Comment by kanielrkirby** [comment]
Ohhhh, yeah I ran the wrong command, that's the flake command. It's supposed to be `nix-build` not `nix build`. One second.

**Comment by MilesCranmer** [comment]
My question is more so: if there were any typos in any of SHAs or URLs over all of the nix files, would I have a way of knowing? (Since I’m new to nix, I think it’s good if the test coverage is complete here)

Maybe all the files are already used by that test but just want to make sure. (I thought the default.nix might be for something else)

**Comment by MilesCranmer** [comment]
Oh I think the issue was the CI was executing `rip2` but the binary name is actually just `rip`. Fixed it

**Comment by koalazub** [comment]
> My question is more so: if there were any typos in any of SHAs or URLs over all of the nix files, would I have a way of knowing? (Since I’m new to nix, I think it’s good if the test coverage is complete here)

If the SHA is incorrect it'll fail with a 

> Wanted
> Got 

Kind of error. It won't actually succeed if the SHA is incorrect

**Comment by MilesCranmer** [comment]
Awesome, thanks

**Review by MilesCranmer — APPROVED** [review]

**Comment by kanielrkirby** [comment]
Awesome! It's official! Lol

```bash
❯ nix shell "github:MilesCranmer/rip2"
❯ rip --version
rip 0.8.0
```

**Comment by MilesCranmer** [comment]
Awesome! Thanks for the PR
